apps/rush-lib/src/cli/utilities/pnpm/PnpmLinkManager.ts

```diff
@@ -0,0 +1,190 @@
+
+    const commonPackage: BasePackage = BasePackage.createVirtualTempPackage(packageJsonFilename, installFolderName);
+
+    const localPackage: BasePackage = BasePackage.createLinkedPackage(
+      project.packageJson.name,
+      commonPackage.version,
+      project.projectFolder,
+      commonPackage
+    );
+
+    // now that we have the temp package.json, we can go ahead and link up all the direct dependencies
+
+    // first, start with the rush dependencies, we just need to link to the project folder
+    Object.keys(commonPackage.packageJson!.rushDependencies || {}).forEach((dependencyName: string) => {
+
+      const matchedRushPackage: RushConfigurationProject | undefined =
+        this._rushConfiguration.getProjectByName(dependencyName);
+
+      if (matchedRushPackage) {
+        const matchedVersion: string = matchedRushPackage.packageJson.version;
+
+        let localLinks: string[] = rushLinkJson.localLinks[localPackage.name];
+        if (!localLinks) {
+          localLinks = [];
+          rushLinkJson.localLinks[localPackage.name] = localLinks;
+        }
+        localLinks.push(dependencyName);
+
+        // We did not find a suitable match, so place a new local package that
+        // symlinks to the Rush project
+        const newLocalFolderPath: string = path.join(localPackage.folderPath, 'node_modules', dependencyName);
+
+        const newLocalPackage: BasePackage = BasePackage.createLinkedPackage(
+          dependencyName,
+          matchedVersion,
+          newLocalFolderPath
+        );
+
+        newLocalPackage.symlinkTargetFolderPath = matchedRushPackage.projectFolder;
+        localPackage.children.push(newLocalPackage);
+      } else {
+        // weird state or program bug
+        throw Error('Cannot find rush dependency in rush configuration');
+      }
+    });
+
+    // Iterate through all the regular dependencies
+
+    // With NPM, it's possible for two different projects to have dependencies on
+    // the same version of the same library, but end up with different implementations
+    // of that library, if the library is installed twice and with different secondary
+    // dependencies.The NpmLinkManager recursively links dependency folders to try to
+    // honor this. Since PNPM always uses the same physical folder to represent a given
+    // version of a library, we only need to link directly to that folder, and it will
+    // have a consistent set of secondary dependencies.
+
+    // each of these dependencies should be linked in a special folder that PNPM
+    // creates for the installed version of each .TGZ package, all we need to do
+    // is re-use that symlink in order to get linked to whatever PNPM thought was
+    // appropriate. This folder is usually something like:
+    // C:\{path-to-tgz-with-slashes-replaced-with-"%2F"}\node_modules\{package-name}
+
+    const slashEscapeRegExp: RegExp = new RegExp(path.sep.replace('\\', '\\\\'), 'g');
```

Comment:
>const [](start = 4, length = 5)

This code is unreadable

Reply:
Can you just copy PNPM's code for this, if they don't already expose it?

---
In reply to: [153380617](https://github.com/Microsoft/web-build-tools/pull/449#discussion_r153380617) [](ancestors = 153380617)